src/metabase/notification/seed.clj

```diff
@@ -1,124 +1,178 @@
+    (t2/delete! :model/ChannelTemplate :id [:in template-ids])))
 
 (defn- create-notification!
-  [{:keys [notification subscriptions handlers]}]
-  (let [handlers (for [handler handlers]
+  [notification]
+  (let [handlers (for [handler (:handlers notification)]
                    (if-let [template (:template handler)]
                      (-> handler
                          (dissoc :template)
                          (assoc :template_id (t2/insert-returning-pk!
                                               :model/ChannelTemplate
                                               template)))
                      handler))]
-    (models.notification/create-notification! notification subscriptions handlers)))
+    (models.notification/create-notification!
+     (dissoc notification :handlers :subscriptions)
+     (:subscriptions notification)
+     handlers)))
 
-(defn- seed-notification!
-  "Seed default notifications"
-  []
-  (log/info "Seeding default notifications")
-  (doseq [row @default-notifications]
-    (create-notification! row))
-  (log/infof "Seeded %d notifications" (count @default-notifications)))
+(defn- replace-notification!
+  "Replace an existing notification with a new one"
+  [existing-row {:keys [internal_id] :as new-row}]
+  (cleanup-notification! internal_id existing-row)
+  (create-notification! new-row))
+
+(defn- action
+  [existing-row new-row]
+  (cond
+    (nil? existing-row)
+    :create
+    ;; use json to compare to avoid any difference like keyword vs string
+    (not= (json/encode (sanitize-notification existing-row)) (json/encode (sanitize-notification new-row)))
+    :replace
+    :else
+    :skip))
 
-(defn truncate-then-seed-notification!
-  "Truncate all notifications related tables then seed it with the default notifications.
-  Runs on every deployment by metabase.task.seed_notification.
+(defn- hydrate-notification
+  [notification]
+  (t2/hydrate notification
+              :subscriptions
+              [:handlers :channel :template :recipients]))
 
-  We do this instead of migrations because we want to be able to modify the default notifications
-  without having to write migrations.
+(defn- sync-notification!
+  [{:keys [internal_id] :as row}]
+  (let [existing-notification (some-> (t2/select-one :model/Notification :internal_id internal_id)
+                                      hydrate-notification)]
 
-  This is possible because as of now users can't modify notifications or channel template."
+    (u/prog1 (action existing-notification row)
+      (case <>
+        :create
+        (do
+          (log/debugf "Creating notification %s" internal_id)
+          (create-notification! row))
+        :replace
+        (do
+          (log/debugf "Replacing notification %s" internal_id)
+          (replace-notification! existing-notification row))
+        :skip
+        (do
+          (log/debugf "Skipping notification %s" internal_id)
+          nil)))))
```

Comment:
Log already returns nil I think

Reply:
Hmm you don't even seem to use this return value anyway.

Reply:
yea removed the do